core/lensQueryHelpers.ml

```diff
@@ -0,0 +1,470 @@
+    "'" ^ Str.global_replace (Str.regexp "'") "''" str ^ "'"
+  method quote_field f =
+    "`" ^ Str.global_replace (Str.regexp "`") "``" f ^ "`"
+  method! make_insert_returning_query : (string * string list * string list list * string) -> string list =
+    failwith "Dummy database make_insert_returning_query not supported"
+end
+  
+let comma_seperated a b =
+    if a = "" then b else a ^ ", " ^ b
+
+let rec construct_query_db (expr : lens_phrase) (db : Value.database) (mapCol : string -> string) =
+    let construct_query expr = construct_query_db expr db mapCol in
+    match expr with
+    | `Constant c -> Constant.string_of_constant c
+    | `Var v -> mapCol v
+    | `InfixAppl (op, a1, a2) -> construct_query a1 ^ " " ^ translate_op_to_sql (string_of_binop op) ^ " " ^ construct_query a2 
+    | `TupleLit l -> "(" ^ List.fold_left (fun a b -> a ^ ", " ^ (construct_query b)) (construct_query (List.hd l)) (List.tl l)  ^ ")"
+    | `UnaryAppl (op, a1) ->  
+        begin
+            match string_of_unary_op op with
+            | "!" -> "NOT (" ^ construct_query a1 ^ ")"
+            | a -> a ^ " (" ^ construct_query a1 ^ ")"
+        end
+    | `In (names, vals) -> 
+        let b = Buffer.create 255 in
+        let cc = Buffer.add_string b in
+        cc "("; List.iteri (fun i v -> if i > 0 then cc ", "; cc (mapCol v)) names; cc ") IN (";
+        List.iteri (fun i v -> if i > 0 then cc ", "; cc "(";
+            List.iteri (fun i2 v2 -> if i2 > 0 then cc ", "; cc (Constant.string_of_constant v2)) v; cc ")") vals;
+            cc ")";
+        Buffer.contents b
+    | `Case (inp, cases, otherwise) -> 
+            let otherwise = construct_query otherwise in
+            if cases = [] then
+                (* special case: if there are no cases then just return query for otherwise *)
+                otherwise
+            else
+                let cases = List.fold_left (fun a (k,v) -> a ^ " WHEN " ^ construct_query k ^ " THEN " ^ construct_query v) "" cases in
+                (match inp with None -> "CASE" | Some inp -> "CASE (" ^ construct_query inp ^ ")") ^
+                    cases ^ " ELSE " ^ otherwise ^ " END"
+
+let construct_query (expr : lens_phrase) =
+  let db = (new dummy_database) in
+  let mapCol = fun a -> db#quote_field a in
+  construct_query_db expr db mapCol  
+
+let construct_select_query (query : select_query) =
+  let db = query.db in
+  let colFn col = db#quote_field col.table ^ "." ^ db#quote_field col.name ^ " AS " ^ db#quote_field col.alias in 
+  let tableFn (table, alias) = db#quote_field table ^ " AS " ^ db#quote_field alias in
+  let cols = List.filter (fun c -> c.present) query.cols in
+  let cols = List.fold_left (fun a b -> a ^ ", " ^ colFn b) (colFn (List.hd cols)) (List.tl cols) in
+  let tables = query.tables in
+  let tables = List.fold_left (fun a b -> a ^ ", " ^ tableFn b) (tableFn (List.hd tables)) (List.tl tables) in
+  let sql = "SELECT " ^ cols ^ " FROM " ^ tables in
```

Comment:
Should this be SELECT DISTINCT given that you want set-valued semantics?  Or does this never matter because the queries always return (at least) a key?